tap_mailchimp/streams.py

```diff
@@ -0,0 +1,545 @@
+import json
+import os
+import time
+import random
+import tarfile
+
+import singer
+from singer import metrics, metadata, Transformer
+from singer.utils import strptime_to_utc, should_sync_field
+from requests.exceptions import HTTPError
+
+LOGGER = singer.get_logger()
+
+MIN_RETRY_INTERVAL = 2 # 2 seconds
+MAX_RETRY_INTERVAL = 300 # 5 minutes
+MAX_RETRY_ELAPSED_TIME = 43200 # 12 hours
+
+# Break up reports_email_activity batches to iterate over chunks
+EMAIL_ACTIVITY_BATCH_SIZE = 100
+
+class BatchExpiredError(Exception):
+    pass
+
+def next_sleep_interval(previous_sleep_interval):
+    min_interval = previous_sleep_interval or MIN_RETRY_INTERVAL
+    max_interval = previous_sleep_interval * 2 or MIN_RETRY_INTERVAL
+    return min(MAX_RETRY_INTERVAL, random.randint(min_interval, max_interval))
+
+def get_abs_path(path):
+    return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)
+
+def chunk_campaigns(sorted_campaigns, chunk_bookmark):
```

Comment:
Add function doc string for all the functions.

Reply:
Added doc string for all functions.